src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs

```diff
@@ -1149,18 +1150,34 @@ public static IEnumerable<Diagnostic> GetEffectiveDiagnostics(ImmutableArray<Dia
 
         private static IEnumerable<Diagnostic> GetEffectiveDiagnosticsImpl(ImmutableArray<Diagnostic> diagnostics, Compilation compilation)
         {
+            if (diagnostics.IsEmpty)
+            {
+                yield break;
+            }
+
             var suppressMessageState = new SuppressMessageAttributeState(compilation);
+            var semanticModelsByTree = new ConcurrentDictionary<SyntaxTree, SemanticModel>();
             foreach (var diagnostic in diagnostics)
             {
                 if (diagnostic != null)
                 {
                     var effectiveDiagnostic = compilation.Options.FilterDiagnostic(diagnostic);
                     if (effectiveDiagnostic != null)
                     {
-                        yield return suppressMessageState.ApplySourceSuppressions(effectiveDiagnostic);
+                        yield return suppressMessageState.ApplySourceSuppressions(effectiveDiagnostic, getSemanticModel);
                     }
                 }
             }
+
+            SemanticModel getSemanticModel(SyntaxTree tree)
+            {
+                if (!semanticModelsByTree.TryGetValue(tree, out var model))
+                {
+                    model = semanticModelsByTree.GetOrAdd(tree, compilation.GetSemanticModel(tree));
+                }
```

Comment:
`return semanticModelsByTree.GetOrAdd(tree, tree => ...);`

Reply:
That will end up creating a delegate for every query, even when the entry exists in the map.